### PR TITLE
GTH-692: Don't add collateral when not needed

### LINKED ISSUE #692: Contract always includes collateral input
### Summary

Contract includes collateral input when there are no plutus script being executed in transaction. It should only include collateral input when needed, unless overridden by `mustUseOutputAsCollateral` constraint.

### Steps to reproduce the behavior

1. Run any of the plutus-contract tests to fail (e.g. change validated transaction count)
2. Observe logs for the initial transaction sending funds to script address

### Actual Result

The "Balancing an unbalanced transaction" does not have any collateral inputs but the "Finished balancing" does show one.

### Expected Result

No collateral inputs are used for a regular payment where no plutus scripts are being executed.

### Describe the approach you would take to fix this

_No response_

### System info

plutus-apps `main` branch f592a7b2a97e3e2a318f5585661c5ada1fe9bd43

### PR DESCRIPTION
**Pull request by sjoerdvisscher**
Fixes #692 

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Formatting, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [ ] Reference the ADR in the PR and reference the PR in the ADR (if revelant)
    - [x] Reviewer requested

### CODE REVIEW

File: plutus-contract/src/Wallet/Emulator/Wallet.hs [hunk 1]
```diff
@@ -459,22 +459,30 @@ handleBalanceTx utxo utx = do
...
-            logDebug NoCollateralInputsAdded
-            pure txWithinputsAdded
-        else do
-            logDebug $ AddingCollateralInputsFor negColl
-            pure $ txWithinputsAdded & over Tx.collateralInputs (sort . (++) (fmap Tx.pubKeyTxInput newTxInsColl))
-
-    pure $ txWithCollateralInputs & Tx.totalCollateral ?~ collFees & Tx.returnCollateral .~ mNewTxOutColl
+    if Value.isZero (fold collateral)
+        && null (Tx.txScripts txWithinputsAdded)
+        && null (Tx.txReferenceInputs txWithinputsAdded)
```
Comment (koslambrou) [inline plutus-contract/src/Wallet/Emulator/Wallet.hs:464]: I guess you _can_ reference a txout that _doesn't_ have an reference script. So this condition needs to be a bit different?

File: plutus-contract/src/Wallet/Emulator/Wallet.hs [hunk 2]
```diff
@@ -459,22 +459,30 @@ handleBalanceTx utxo utx = do
...
-        then do
-            logDebug NoCollateralInputsAdded
-            pure txWithinputsAdded
-        else do
-            logDebug $ AddingCollateralInputsFor negColl
-            pure $ txWithinputsAdded & over Tx.collateralInputs (sort . (++) (fmap Tx.pubKeyTxInput newTxInsColl))
-
-    pure $ txWithCollateralInputs & Tx.totalCollateral ?~ collFees & Tx.returnCollateral .~ mNewTxOutColl
+    if Value.isZero (fold collateral)
+        && null (Tx.txScripts txWithinputsAdded)
```
Comment (koslambrou) [inline plutus-contract/src/Wallet/Emulator/Wallet.hs:463]: I think we also need to add:

```
&& null (Tx.txMintingScripts txWithinputsAdded)
```

File: plutus-use-cases/test/Spec/crowdfundingEmulatorTestOutput.txt [hunk 3]
```diff
@@ -40,28 +40,22 @@ Slot 1: W[2]: Balancing an unbalanced transaction:
...
-                    - Value (Map [(,Map [("",9815623)])]) addressed to
-                      PubKeyCredential: 80a4f45b56b88d1139da23bc4c3c75ec6d32943c087f250b86193ca7 (no staking credential)
-                  return collateral:
-                    - Value (Map [(,Map [("",9723434)])]) addressed to
+                    - Value (Map [(,Map [("",9823763)])]) addressed to
                       PubKeyCredential: 80a4f45b56b88d1139da23bc4c3c75ec6d32943c087f250b86193ca7 (no staking credential)
-                  total collateral: Value (Map [(,Map [("",276566)])])
                   mint: Value (Map [])
-                  fee: Value (Map [(,Map [("",184377)])])
+                  fee: Value (Map [(,Map [("",176237)])])
```
Comment (koslambrou) [inline plutus-use-cases/test/Spec/crowdfundingEmulatorTestOutput.txt:58]: :+1:

File: plutus-contract/src/Wallet/Emulator/Wallet.hs (hunk 2 above)
Comment (sjoerdvisscher) [reply plutus-contract/src/Wallet/Emulator/Wallet.hs:463]: `txMintingScripts` only contains redeemers, the actual scripts are still in `txScripts`.

File: plutus-contract/test/Spec/TxConstraints/MustSpendPubKeyOutput.hs [hunk 4]
```diff
@@ -69,10 +69,13 @@ baseLovelaceLockedByScript :: Integer
...
-mustSpendPubKeyOutputContract offChainTxOutRefs onChainTxOutRefs pkh = do
+mustSpendPubKeyOutputContract = mustSpendPubKeyOutputContract' []
+
+mustSpendPubKeyOutputContract' :: [Ledger.PaymentPubKeyHash] -> [TxOutRef] -> [TxOutRef] -> Ledger.PaymentPubKeyHash -> Contract () Empty ContractError ()
+mustSpendPubKeyOutputContract' keys offChainTxOutRefs onChainTxOutRefs pkh = do
     let lookups1 = Constraints.typedValidatorLookups typedValidator
-        tx1 = Constraints.mustPayToTheScriptWithDatumInTx onChainTxOutRefs
-            $ Ada.lovelaceValueOf baseLovelaceLockedByScript
+        tx1 = Constraints.mustPayToTheScriptWithDatumInTx onChainTxOutRefs (Ada.lovelaceValueOf baseLovelaceLockedByScript)
+            <> foldMap Constraints.mustBeSignedBy keys
```
Comment (james-iohk) [inline plutus-contract/test/Spec/TxConstraints/MustSpendPubKeyOutput.hs:78]: Why is this now needed?

Comment (james-iohk) [reply plutus-contract/test/Spec/TxConstraints/MustSpendPubKeyOutput.hs:78]: That's possible. Not to worry. Thanks!

**Review by james-iohk — APPROVED** [review]

File: plutus-contract/test/Spec/TxConstraints/MustSpendPubKeyOutput.hs (hunk 4 above)
Comment (james-iohk) [reply plutus-contract/test/Spec/TxConstraints/MustSpendPubKeyOutput.hs:78]: Actually, something to note is that `mustBeSignedBy` includes required signer in the txbody, which is redundant in the case where onchain validation for this witness is not performed.

**Pull request merged**